File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private readonly int _speed;
    private readonly int _batteryDrain;

    private int metersDriven;
    private int batteryPercentage = 100;

    public RemoteControlCar(int speed, int batteryDrain) 
    {
        _speed = speed;
        _batteryDrain = batteryDrain;
    }

    public bool BatteryDrained()
    {
        return batteryPercentage < _batteryDrain;
    }

    public int DistanceDriven()
    {
        return metersDriven;
    }

    public void Drive()
    {
        if (BatteryDrained())
            return;
        
        metersDriven += _speed;
        batteryPercentage -= _batteryDrain;
    }

    public static RemoteControlCar Nitro()
    {
        return new RemoteControlCar(50, 4);
    }
}

class RaceTrack
{
    private readonly int _distanceInMeters;
    
    public RaceTrack(int distanceInMeters) 
    {
        _distanceInMeters = distanceInMeters;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        if (car.DistanceDriven() == _distanceInMeters)
            return true;

        if (car.BatteryDrained())
            return false;

        car.Drive();
        return TryFinishTrack(car);
    }
}

File: solutions/java/need-for-speed/2/src/main/java/NeedForSpeed.java
public final class NeedForSpeed {
    private final int speed;
    private final int batteryDrain;
    
    private int battery = 100;
    private int distanceDriven = 0;
    
    public NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return battery < batteryDrain;
    }

    public int distanceDriven() {
        return distanceDriven;
    }

    public void drive() {
        if (!batteryDrained()) {
            distanceDriven += speed;
            battery -= batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

final class RaceTrack {
    private int distance;
    
    public RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        while (!car.batteryDrained())
            car.drive();

        return distance <= car.distanceDriven();
    }
}
